[MO] - [blog post] -> how system test works? (#186)
strimzi/strimzi.github.io: _posts/2020-09-06-how-the-system-tests-works.md

    super.tearDownEnvironmentAfterAll();
}
```

In order to delete all resources from specific `Resources` instance, execute:
```
ResourceManager.deleteMethodResources();
ResourceManager.deleteClassResources();
```

## Resources

Resources are in our case an abstraction of the `.yaml` definitions of custom resources provided by Strimzi, such as 
`Kafka`, `KafkaTopic`, `KafkaUser`, `KafkaConnect`, `KafkaMirrorMaker` and so on.
These custom resources are represented as Java objects. 
This yaml representation of a `KafkaTopic`

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: KafkaTopic
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec:
  partitions: 1
  replicas: 1
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824
```

can be also be represented as the following Java code:

```
KafkaTopic kafkaTopic = getKafkaTopicFromYaml(PATH_TO_KAFKA_TOPIC_CONFIG); 
// additional changes...
kafkaTopic = new KafkaTopicBuilder(kafkaTopic)
    .withNewMetadata()
        .withName(topicName)
        .withNamespace(ResourceManager.kubeClient().getNamespace())
        .addToLabels(Labels.STRIMZI_CLUSTER_LABEL, clusterName)
    .endMetadata()

>     .editSpec()

Reviewer: Either spec/config is wrong or java code editSpec() - as they do not correlate:
    retention.ms: 7200000
    segment.bytes: 1073741824
vs
        .addToConfig("min.insync.replicas", minIsr)

        .withPartitions(partitions)
        .withReplicas(replicas)
        .addToConfig("min.insync.replicas", minIsr)
    .endSpec();
```

Everything is built within [three stacks](https://github.com/strimzi/strimzi-kafka-operator/blob/master/systemtest/src/main/java/io/strimzi/systemtest/resources/ResourceManager.java#L77-L79). 
These three stacks are managed by `ResourceManager`, which is taking care of switching. 
The method stack holds all the resources needed for each the test case. 
It gets cleaned up after each `@Test` method execution.
The class stack holds all the resources for the text fixture.
These are prepared before the test, using the `@BeforeAll`-annotated methods.
The third stack is used as a reference to either of the other two stacks to delete the resources in specific test phases.

> This stack points to the class stack in we are in the `@BeforeAll` scope, otherwise we are working with the method stack. 

Reviewer: `class stack in we are in the` ?

> The logic inside these stacks are that once you create some resource (for instance a `KafkaTopic`),

Reviewer: ```suggestion

The logic inside these stacks are that once you create some resource (for instance a `KafkaTopic`),
```

it will be pushed inside the stack and the deletion of topic will be performed in the teardown phase. 
Once the teardown phase is initiated, the method stack will pop all resources after test case is over. 
The same logic applies with the class stack, which pops and deletes resources when test suite is over. 
By setting the `SKIP_TEARDOWN` environment variable to `TRUE` it will skip the teardown 
phase and all created resources will remain in the stack. This can be very useful for debugging purposes.

## Auxiliary classes

In system tests there are many helper classes, each having a different purpose:

1. `Utils`
2. `Constants`
3. `Environment`
4. Kubernetes client
5. Kafka clients

#### `Utils` classes

We have separate utils for each resource. 
For example, we have `KafkaUtils` for `Kafka` resource, `KafkaTopic` for KafkaResource and so on. Furthermore, we have a number of
Kubernetes-related utils such as `ServiceUtils`, `PodUtils`, `DeploymentsUtils` and so on.
These classes come handy when you have to wait for some change (e.g. wait for a `Pod` to become ready) or create, change or delete a Kubernetes resource.

> The one method which stands out is `waitFor`, which is used in many utils methods.

Reviewer: ```suggestion

The one method which stands out is `waitFor`, which is used in many utils methods.
```

You can imagine the scenario, where you change a custom resource and then you want to assert the outcome is correct. 
In this scenario, if we didn't `waitFor()` the outcome to become correct, the assertion would execute too quickly and the verification would fail. 

> In other cases we the operator needs to perform a rolling update, which takes time, so the test needs to wait until all the pods have been restarted.

Reviewer: ```suggestion
In other cases the operator needs to perform a rolling update, which takes time, so the test needs to wait until all the pods have been restarted.
```

---
layout: post
title:  "How system tests works"
date: 2020-09-06
author: maros_orsak
---
In [previous blog post](https://strimzi.io/blog/2020/09/21/introduction-to-system-tests/) related on testing we learned 
about our unit tests, integration tests and how they differ from system tests. 
Furthermore, we showed up how to run them on Kubernetes cluster. 
But, we did not cover whole understanding of the system tests. 
In this blog post we will take a closer look at our system tests. The content is as follows:

# Content
1. [Introduction](#introduction)
2. [Lifecycle of tests](#lifecycle)
3. [Resources](#resources)
4. [Auxiliary classes](#auxiliaryclasses)
5. [Dependencies](#dependencies)
6. [How to create a system test](#createtest)

> 8. [Conclusion](#conclusion)

Author: I have removed it.

## <a id="introduction">Introduction</a>

The whole ecosystem behind systems test is encapsulated in two mandatory things. 
First the `Resources`, creates the whole testing environment. 
Second the auxiliary classes, which are divided into the classic static methods `Utils`, internal clients, 
`Apache Kafka Clients` for the external communication, Kubernetes client, `Constants.java` and `Environment.java`.

> Main idea is to make the system tests and its resources easily modifiable and writable in the fluent way. Let's have a closer

Author: Yes modified.

one for test method resources and one auxiliary, which always points to class or method stack.
You are allowed to create resources during execution of any test method. 
Resource lifecycle implementation handles insertion of the resource on top of the stack and its deletion at the end of the test method/class (teardown phase).

`ResourceManager` stores info, which stack is currently active (class or method) in the `pointer stack`.
You can change between class and method resources stack with method `ResourceManager.setMethodResources()` or `ResourceManager.setClassResources()`.
Note that pointer stack is set automatically in `AbstractST.class` in `@BeforeAll` or `@BeforeEach` methods.

Cluster Operator setup example:

```
@BeforeAll
void createClassResources() {
   ResourceManager.setClassResources();     <--- point resource manager to class stack
   installClusterOperator(NAMESPACE);       <--- install cluster operator
}
```

### Exercise
In this phase we specify all steps which need to be executed to cover some specific functionality. Reasonable example would be
the Code snippet 1.4 from previous section, where we setup Kafka cluster with new persistent claim storage.

> ### Verify

Author: Great note

What is worth to mention, is that if you specify `SKIP_TEARDOWN` environment variable to `TRUE` it will skip the teardown 
phase and all created resources will remain in the stack. Also, mostly great for debugging because it could break tests 
if it's set for multiple than one.

## <a id="auxiliaryclasses">Auxiliary classes</a>

In `system test` module there are many helper classes,  each having a different purpose:

1. Utils
2. Constants
3. Environment
4. Kubernetes client
5. Kafka clients

#### Utils

These classes are applicable on many problems. We have separate utils for each resource. 
For instance we have KafkaUtils for KafkaResource, KafkaTopic for KafkaResource and so on. Furthermore, we have got
Kubernetes related utils such as ServiceUtils, PodUtils, DeploymentsUtils and many more.
These classes come handy for instance when you have to wait for some change, make some API call, create, change or delete a Kubernetes resource and much more.
The one method which stands out is `waitFor`, which is used in many utils methods.
You can imagine the scenario, where you will change some custom resource value of Kafka and then you will want to assert

> that this value was correctly updated. This is a very good example where `race condition` comes into play. This race condition

Reviewer 2: I'm not sure _race condition_ is the right term here. I would maybe more explain how the things happen asynchronously and with rolling updates etc. can take seconds or minutes to happen and that is why the `waitFor` is important and useful.

The path is defined by environment variable `ST_CONFIG_PATH`. If it is not defined, the default config file location `systemtest/config.json` will be used instead.
Loading of system configuration has the following priority order:
1. Environment variable
2. Variable defined in configuration file
3. Default value

Here is the list of important environment variables:

1. `DOCKER_ORG` = Specifies the organization/repo containing the image used in system tests (default value `strimzi`)
2. `DOCKER_TAG` = Specifies the image tags used in system tests (default value `latest`)
3. `SKIP_TEARDOWN` = Skip teardown phase - primarily used for debug purposes (default value `false`)
4. `ST_KAFKA_VERSION` =  Specifies Kafka version used in images during the system tests (default value `2.6.0`)

If you want to see all our environment variables you can see it [here](https://github.com/strimzi/strimzi-kafka-operator/blob/master/development-docs/TESTING.md#environment-variables)

#### Our Kafka clients

Another helpful module of the system tests are Kafka clients, which contains three usage types:

1. For internal communication within Kubernetes - using test-client image

> 2. For internal communication within Kubernetes - client-examples image

Reviewer 2: Might be worth explaining why two kinds of internal clients. Are there cases whn to use one and when to use another? Or is one of them _deprecated_ and should not be used for new tests?

Author: Yes, internal clients which are using the test-client image are deprecated I will mention this one.

                            .endEnv()
                            .addNewEnv()
                                .withName("DELAY_MS")
                                .withValue("1000")
                            .endEnv()
                            .addNewEnv()
                                .withName("LOG_LEVEL")
                                .withValue("DEBUG")
                            .endEnv()
                            .addNewEnv()
                                .withName("MESSAGE_COUNT")
                                .withValue(String.valueOf(messageCount))
                            .endEnv()
                            .addNewEnv()
                                .withName("PRODUCER_ACKS")
                                .withValue("all")
                            .endEnv()
                            .addNewEnv()
                                .withName("ADDITIONAL_CONFIG")
                                .withValue(additionalConfig)
                            .endEnv()
                    .endContainer()
                .endSpec()
            .endTemplate()
        .endSpec()
        .build());
```

##### External using Vert.x client

> Lastly, external clients are implemented by supporting Vertx.io Verticles. This type of client has the most limitations for usage. Most notable is dependency on external listeners, which is caused by`platform dependency`. 

Reviewer 2: Can you explain what the _platform dependency_ is? Loadbalancers, Routes, NodePorts etc.?

> They are implemented by _Abstract recursive builder, which makes these clients very handful. 

Reviewer 2: The `_` in `_Abstract` renders in the text. Was it supposed to be closed somewhere for italic font?

Author: it was just an error removed.

What is worth mentioning is that external clients are running in same environment as your tests.
Moreover, we have `Oauth`, `Tracing` and also `Basic` client. Each of them are for specific use case. 
However, we prefer to always use the internal clients because of mentioned platform dependency. 
Client can be used for instance:

```
// 1. Creating the client instance 

BasicExternalKafkaClient basicExternalKafkaClient = new BasicExternalKafkaClient.Builder()
    .withTopicName("topic-name")
    .withNamespaceName("namespace-name")
    .withClusterName("cluster-name")
    .withKafkaUsername("kafka-username")
    .withMessageCount(100)
    .withSecurityProtocol(SecurityProtocol.SSL)
    .build();

// 2. Sending messages

basicExternalKafkaClient.sendMesssageTls();
```

## <a id="dependencies">Dependencies</a>

Before writing simple test case in Strimzi system tests, we need to know main dependencies that `system test` module has. 
The most important ones are:
1. test (kubernetes client)

> 2. crd-generator + api (chaining / builder pattern)

Reviewer 2: Do you really depend on crd-generator? I would probably just keep the `api` here.

which servers as executor and can invoke various requests such as `kubectl annotate...`, `kubectl exec...` and so on.

### crd-generator + api module

Resources previously described in Resource section are used in crd-generator and api modules.
Every time, when the code of Model changes you have to build the whole modules to generate new schema for resources. 
For instance, imagine that you are using `Kafka` resource and you need to create Kafka with the `Cruise Control` support but you have old schemas, which does not support it.
You will need to build these modules and generate schema, which will support it and be able to create Kafka with Cruise Control. 
These modules provide the fluent way of creating Strimzi custom resource instances such as `Kafka`, `KafkaConnect`, `KafkaConnectS2I`, `KafkaMirrorMaker` and many other.

## <a id="createtest">How to create a system test</a>

At this point you have the necessary knowledge to create a simple test. There are also some additional steps, which need to be performed to set up your testing environment.

### Precursors

> 1. Fork the `strimzi-kafka-operator` repository - https://github.com/strimzi/strimzi-kafka-operator

Reviewer 2: Use link instead of test for the URL.

2. Build the whole project in root directory - `mvn clean install -DskipTests` or you can use `mvn -am clean install -DskipTests -f systemtest/pom.xml`
3. Setup your Kubernetes cluster

> 4. Login to your Kubernetes cluster

Reviewer 2: Should you mentioned the priviledges needed to install Strimzi (not list all of them, but say _Login to your Kubernetes cluster as a user who has the rights to install the Strimzi operator_)

> 5. Try to start some of our tests inside `system test` module for instance - `io.strimzi.systemtest.bridge.HttpBridgeST.testSendMessagesTlsAuthenticated()` 

Reviewer 2: On many places, you talk about `system test` module. But the module name is `systemtest`. You should either use the right module name or or just talk about _system tests_ without specifically referring to the module.

Author: Done and replaced

(test takes around 5 minutes)

### Writing new test case

We are going to create a very simple test using aforementioned classes like `Resources`, `Utils`, `Kubernetes client` and so on.

Description of the test case: 

1. Verify that Kafka was successfully deployed
2. Clients are able to send and receive messages to Kafka cluster
 
As you can see, we have two main things to take care of. First is `verify that Kafka was correctly deployed` and the second one is `clients can send messages to and receive messages from Kafka cluster`.

Test scenario could look like this:

1. Deploy Strimzi Operator <- setup + exercise
2. Deploy `Kafka` cluster with 3 kafka nodes and 3 zookeeper nodes <- setup + exercise
3. Verify that `Kafka` is ready <- verify
4. Create instance of `Kafka clients` <- setup + exercise
5. Send messages to `Kafka` cluster <- exercise
6. Receive messages from `Kafka` cluster <- exercise
7. Verify that we have sent and received expected count of messages <-verify

> 8. Clean up all resources (this phase is implicitly done in AbstractST) <- teardown

Reviewer 2: The `<-` are quite hard to read. I think you should try to rephrase the points to avoid it.

Note, that teardown phase is performed automatically by system tests mechanism so there is no need to take care of it by user in code.

Test case then can look like this:

```java
package io.strimzi.systemtest;

import io.strimzi.systemtest.kafkaclients.internalClients.InternalKafkaClient;
import io.strimzi.systemtest.resources.ResourceManager;
import io.strimzi.systemtest.resources.crd.KafkaClientsResource;
import io.strimzi.systemtest.resources.crd.KafkaResource;
import io.strimzi.systemtest.utils.kafkaUtils.KafkaUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

public class MyTestSuite extends AbstractST {

    private static final String NAMESPACE = "my-test-suite-namespace";
    private static final Logger LOGGER = LogManager.getLogger(MyTestSuite.class);

    @Test

>     void myFirstTestCase() {

Reviewer 2: Should it have some tag annotation?

Author: yeah maybe `@TAG(INTERNAL_CLIENTS_USED)`

---
layout: post
title:  "How system tests work"
date: 2020-09-06
author: maros_orsak
---
In [previous blog post](https://strimzi.io/blog/2020/09/21/introduction-to-system-tests/) related on testing we learned 
about our unit tests, integration tests and how they differ from system tests. 
Furthermore, we showed up how to run them on Kubernetes cluster. 

> But, we did not cover whole understanding of the system tests. 

Author: What do you thing about this one? `But, we did not cover the system tests themselves in any great detail.`

> But, we did not cover whole understanding of the system tests. 

Reviewer 3: Good spot: We should drop the comma

In this blog post we will take a closer look at our system tests. The content is as follows:

## Introduction

> The whole ecosystem behind systems test is encapsulated in two mandatory things. 

Reviewer 4: `system tests`?

You can change between class and method resources stack with method `ResourceManager.setMethodResources()` or `ResourceManager.setClassResources()`.
Note that pointer stack is set automatically in `AbstractST.class` in `@BeforeAll` or `@BeforeEach` methods.

Cluster Operator setup example:

```
@BeforeAll
void createClassResources() {
   ResourceManager.setClassResources();     <--- point resource manager to class stack
   installClusterOperator(NAMESPACE);       <--- install cluster operator
}
```

### Exercise
In this phase we prepare our component, which is tested. In our case it is set of operation that can modified state of our 
components for instance updating configuration of `Kafka`, `KafkaConnect` or creating new listeners in `Kafka` and more.

### Verify

When your environment is ready from the previous phase then incomes checks of state, message exchange and so on.
In this code snippet you can see verification phase, where we verify that configuration of the Kafka custom resource is successfully set.
Moreover, we verify that specific property is correctly set inside Kafka pod.

> ```

Reviewer 4: Does it makes sense to add the ```java for the all other code blocks?

Author: I have tried to have all java code blocks but in our cases, on website it is not pretty.

Reviewer 4: Ah okay, well you have there one java code block, so in case it's not pretty, can you please delete it?

The path is defined by environment variable `ST_CONFIG_PATH`. If it is not defined, the default config file location `systemtest/config.json` will be used instead.
Loading of system configuration has the following priority order:
1. Environment variable
2. Variable defined in configuration file
3. Default value

Here is the list of important environment variables:

1. `DOCKER_ORG` = Specifies the organization/repo containing the image used in system tests (default value `strimzi`)
2. `DOCKER_TAG` = Specifies the image tags used in system tests (default value `latest`)
3. `SKIP_TEARDOWN` = Skip teardown phase - primarily used for debug purposes (default value `false`)
4. `ST_KAFKA_VERSION` =  Specifies Kafka version used in images during the system tests (default value `2.6.0`)

If you want to see all our environment variables you can see it [here](https://github.com/strimzi/strimzi-kafka-operator/blob/master/development-docs/TESTING.md#environment-variables)

#### Our Kafka clients

Another helpful module of the system tests are Kafka clients, which contains three usage types:

1. For internal communication within Kubernetes - client-examples image

> 2. For external communication within Kubernetes - using Vert.x client

Reviewer 2: Add link to Vert.x website please.

### Precursors

1. Fork the [strimzi-kafka-operator](https://github.com/strimzi/strimzi-kafka-operator) repository
2. Build the whole project in root directory - `mvn clean install -DskipTests` or you can use `mvn -am clean install -DskipTests -f systemtest/pom.xml`
3. Setup your Kubernetes cluster
4. Login to your Kubernetes cluster as a user who has the rights to install the Strimzi operator
5. Try to start some of our tests inside system tests for instance - `io.strimzi.systemtest.bridge.HttpBridgeST.testSendMessagesTlsAuthenticated()` 
(test takes around 5 minutes)

### Writing new test case

We are going to create a very simple test using aforementioned classes like `Resources`, `Utils`, `Kubernetes client` and so on.

Description of the test case: 

1. Verify that Kafka was successfully deployed
2. Clients are able to send and receive messages to Kafka cluster
 
As you can see, we have two main things to take care of. First is `verify that Kafka was correctly deployed` and the second one is `clients can send messages to and receive messages from Kafka cluster`.

Test scenario could look like this:

> 1. Setup & Exercise:  Deploy `Strimzi Operator

Reviewer 4: ```suggestion
1. Setup & Exercise:  Deploy `Strimzi Operator`
```
This will "code block" will continue to next lines and "destroy" the "final look"

Author: Good catch :)

---
layout: post
title:  "How system tests works"
date: 2020-09-06
author: maros_orsak
---
In [previous blog post](https://strimzi.io/blog/2020/09/21/introduction-to-system-tests/) related on testing we learned about our unit tests, integration tests and how they differ from system tests. 

> Furthermore, we showed up how to run them against Kubernetes cluster. But, we did not covered whole understanding of the

Reviewer: on Kubernetes cluster?

> system tests. Today, we will take a closer look on our system tests. The content is as follows:

Reviewer: ```suggestion
system tests. Today, we will take a closer look at our system tests. The content is as follows:
```

# Content
1. [Introduction](#Introduction)
2. [Resources](#Resources)
3. [Lifecycle of tests](#Lifecycle of tests)
4. [Auxiliary classes](#Auxiliary classes)
5. [Dependencies](#Dependencies)
6. [How to create a system test](#How to create a system test)
8. [Conclusion](#Conclusion)

## Introduction

> The whole ecosystem behind systems test is encapsulated in two mandatory things. First of all it is

Reviewer: ```suggestion
The whole ecosystem behind systems test is encapsulated in two mandatory things. First is
```

the `Resources`, which create the whole testing environment. The second of all we have the auxiliary classes that are divided into
the classic static methods `Utils`, internal clients and also `Apache Kafka Clients` for the external communication, 
Kubernetes client, `Constants.java` and `Environment.java`.

> Pure idea to make the system tests easily modifiable is to able to write resources in the fluent way. 

Reviewer: ```suggestion
Main idea was to make the system tests and its resources easily modifiable and writable in the fluent way. 
```

> Fluent notation are nested builders by which you can build customized objects by your imagination.

Reviewer: ```suggestion
Fluent notation is a set of nested builders, which can build customized objects by your needs.
```

For instance before the Fluent notation we had the [1.Telescoping pattern](https://www.vojtechruzicka.com/avoid-telescoping-constructor-pattern/#telescoping-constructor) 

> but this does not scale well for instance in situation with large numbers of optional parameters. 

Reviewer: ```suggestion
but that did not scale well (for instance in situations with large numbers of optional parameters). 
```

Second alternative would be the [2.JavaBeans convention](https://www.vojtechruzicka.com/avoid-telescoping-constructor-pattern/#alternative-1---javabeans), 

> which first of all resolve our issue from the telescoping pattern but on the other hand the objects which we create will be mutable. 

Reviewer: ```suggestion
which could resolve our issue from the telescoping pattern, but on the other hand objects which created objects will be mutable. 
```

Luckily, there is a third alternative that combines the safety of the telescoping constructor pattern with the readability 
of the `JavaBeans pattern`. 
It is a form of the [3.Builder pattern](https://refactoring.guru/design-patterns/builder).
Moreover, if we take this idea to the next level to the `4.Abstract Recursive builder` we will be able to create multiple 
dependent objects, which have nested hierarchy. In the following code snippets [1.1](#javabeans), [1.2](#builder), [1.3](#abstractbuilder) 
we can see described idea and you can see how easily we can create an object using `Abstract Recursive builder`.

####  1. Kafka with new persistent claim storage and following settings

##### <a id="javabeans">Code snippet 1.1 Using JavaBeans convention</a>

```java
PersistentClaimStorage psc = new PersistentClaimStorage();

// JavaBeanss convention having setters 
psc.setId(0);
psc.setSize("100Gi");
psc.setDeleteClaim(true);

KafkaResource.kafkaEphemeral(CLUSTER_NAME, 3, 1)
    .editSpec()
        .editKafka()
            .withPersistentClaimStorage(psc)
        .endKafka()
    .endSpec()
    .done();
```

##### <a id="builder">Code snippet 1.2 Using Builder pattern</a>
```java

> // Builder pattern by which we can create one type of object

Reviewer: ```suggestion
// Builder pattern which creates one type of object
```

PersistentClaimStorage pcs = new PersistentClaimStorageBuilder()
    .withDeleteClaim(true)
    .withId(0)
    .withSize("100Gi")
    .build();

KafkaResource.kafkaEphemeral(CLUSTER_NAME, 3, 1)
    .editSpec()
        .editKafka()
            .withPersistentClaimStorage(psc)
        .endKafka()
    .endSpec()
    .done();
```

##### <a id="abstractbuilder">Code snippet 1.3 Using Abstract recursive builder</a>
```java
// Abstract recursive builder which can have more objects in hierarchy as you can see here
KafkaResource.kafkaEphemeral(CLUSTER_NAME, 3, 1)
    .editSpec()
        .editKafka()
            .withNewPersistentClaimStorage()
                .withId(0)
                .withNewSize("100Gi")
                .withDeleteClaim(true)
            .endPersistentClaimStorage()
        .endKafka()
    .endSpec()
    .done();
```

> This examples was illustrated idea only one dependent you can imagine three, four and more objects, which can be changed for instance:

Reviewer: ```suggestion
These examples illustrate idea of only one dependent. You can change more dependent objects like in following example:
```

```
KafkaResource.kafkaEphemeral(CLUSTER_NAME, 3, 1)
    .editMetadata()
        .withName(name)
        .withNamespace(ResourceManager.kubeClient().getNamespace())
    .endMetadata()
    .editSpec()
        .editKafka()
            .withVersion(Environment.ST_KAFKA_VERSION)
            .withReplicas(kafkaReplicas)
            .addToConfig("log.message.format.version", TestKafkaVersion.getKafkaVersionsInMap().get(Environment.ST_KAFKA_VERSION).protocolVersion())
            .addToConfig("offsets.topic.replication.factor", Math.min(kafkaReplicas, 3))
            .addToConfig("transaction.state.log.min.isr", Math.min(kafkaReplicas, 2))
            .addToConfig("transaction.state.log.replication.factor", Math.min(kafkaReplicas, 3))
            .withNewListeners()
                .withNewPlain().endPlain()
                .withNewTls().endTls()

>             .endListeners()

Author: Yes of course

            .withNewInlineLogging()
                .addToLoggers("log4j.rootLogger", "DEBUG")
            .endInlineLogging()
        .endKafka()
        ...
        ... (more configs...)
        ... 
    .endSpec()
    .done()
```

> In this case you can imagine that it is really easy to create multiple different instances of the `Kafka` instead of having separate objects.

Reviewer: I don't want to imagine things. I already see them in examples :)

> In this case you can imagine that it is really easy to create multiple different instances of the `Kafka` instead of having separate objects.

Reviewer: ```suggestion
In this case you can see that it is really easy to create multiple different instances of the `Kafka` instead of having separate objects.
```

## Resources

> Resources in our way are abstraction to the `.yaml` definitions of custom resources which we provide such as 

Reviewer: ```suggestion
Resources in our mean of usage are abstractions of the `.yaml` definition files of custom resources, which we provide, such as 
```

`Kafka`, `KafkaTopic`, `KafkaUser`, `KafkaConnect`, `KafkaMirrorMaker` and so on.

> These custom resources are encapsulated into java objects and some of you can remind the [ORM(Object Relation Mapping)](https://www.tutorialspoint.com/hibernate/orm_overview.htm) 

Reviewer: ```suggestion
These custom resources are encapsulated into java objects and remind the [ORM(Object Relation Mapping)](https://www.tutorialspoint.com/hibernate/orm_overview.htm) 
```

from databases. 

> In the following Figure [2.1](#kafkatopic) we can see YAML representation of `KafkaTopic` and on the Code snippet 

Reviewer: ```suggestion
In the following yaml code[2.1](#kafkatopic) we can see YAML representation of `KafkaTopic` and on the Code snippet 
```

[2.2](kafkaimplementation) related java object.

> ##### <a id="kafkatopic">Figure 2.1 YAML representation of `KafkaTopic`</a>

Reviewer: ```suggestion
##### <a id="kafkatopic">Code snippet 2.1 YAML representation of `KafkaTopic`</a>
```

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: KafkaTopic
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec:
  partitions: 1
  replicas: 1
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824
```

> ##### <a id="kafkaimplementation">Code snippet 2.2 Loading the YAML file to JAVA representation object.</a>

Reviewer: ```suggestion
##### <a id="kafkaimplementation">Code snippet 2.2 Loading of the YAML file to JAVA representation object.</a>
```

```java
KafkaTopic kafkaTopic = getKafkaTopicFromYaml(PATH_TO_KAFKA_TOPIC_CONFIG); 
// additional changes...
kafkaTopic = new KafkaTopicBuilder(kafkaTopic)
    .withNewMetadata()
        .withName(topicName)
        .withNamespace(ResourceManager.kubeClient().getNamespace())
        .addToLabels(Labels.STRIMZI_CLUSTER_LABEL, clusterName)
    .endMetadata()
    .editSpec()
        .withPartitions(partitions)
        .withReplicas(replicas)
        .addToConfig("min.insync.replicas", minIsr)
    .endSpec();
```

> Everything is build with three stacks. These three stacks are managed by `ResourceManager`, which is taking care of switching. 

Reviewer: ```suggestion
Everything is built within three stacks. These three stacks are managed by `ResourceManager`, which is taking care of switching. 
```

> First one is `method stack` where his responsibility is to hold all resources, which are invoked in the test cases. 

Reviewer: ```suggestion
First one's, `method stack`, responsibility is to hold all resources, which are invoked in the test cases. 
```

> Second one is the `class stack` where his responsibility is to hold all resources, which are invoked in the `@BeforeAll` notation. 

Reviewer: ```suggestion
Second one's, the `class stack`, responsibility is to hold all resources, which are invoked in the `@BeforeAll` notation. 
```

> Switching between these two stacks are doing the third stack called pointer stack. 

Reviewer: ```suggestion
Switching between these two stacks is performed by the third stack called `pointer stack`. 
```

> The pointer stack points to class stack where it is in the `@BeforeAll` notation otherwise we are working with the method stack. 

Reviewer: ```suggestion
The `pointer stack` points to the `class stack` when it is defined in the `@BeforeAll` notation, otherwise we are working with the `method stack`. 
```
Not fully understand this. Hard to correct sentence w/o full meaning..

> The pointer stack points to class stack where it is in the `@BeforeAll` notation otherwise we are working with the method stack. 

Author: Yes, we are always working with pointer stack. This sentencee is little bit complicated.

```
The `pointer stack` points to the `class stack` in we are in the `@BeforeAll` method, otherwise we are working with the `method stack`. 
```
Does this sentence make more sense to you? :)

> The logic inside these stacks are that if you create some resource for instance KafkaTopic. 

Reviewer: ```suggestion
The logic inside these stacks are that once you create some resource (for instance KafkaTopic),
```

> It will be pushed inside the stack and the deletion of topic will be planned. 

Reviewer: ```suggestion
it will be pushed inside the stack and the deletion of topic will be planned in teardown phase. 
```

```

## Auxiliary classes

In our `system test` module we have many classes, which providing help. Each of them has different purpose and here is the list:

1. Utils
2. Constants
3. Environment
4. Kubernetes client
5. Kafka clients

#### Utils

These classes are applicable on many problems. We have separate utils for each resource. 
For instance we have KafkaUtils for KafkaResource, KafkaTopic for KafkaResource and so on. Furthermore, we have got
Kubernetes related utils such as ServiceUtils, PodUtils, DeploymentsUtils and many more. The hierarchy is as follows:

* utils
  * kafkaUtils
    * KafkaBridgeUtils
    * KafkaConnectS2IUtils
    * ...
  * kubeUtils
    * controllers
      * DeploymentConfigUtils
      * DeploymentUtils
      * ...
    * objects
      * NamespaceUtils
      * PersistentVolumeClaimUtils
      * ...
  * specific
    * BridgeUtils
    * CruiseControlUtils
    * ...

The usability of these classes is for instance when you have to wait for some change, calling API, creating, changing, deleting Kubernetes resource and much more.
The one method which deserves to be described is `waitFor`, where you will find it in many utils methods.

> You can imagine the scenario, where you will change some custom resource value of Kafka and then you will want to assert
> that this value was updated but here is very good example where comes `race condition` into play. This race condition

Reviewer: ```suggestion
You can imagine the scenario, where you will change some custom resource value of Kafka and then you will want to assert
that this value was correctly updated. This is a very good example where `race condition` comes into play. This race condition
```

> can occur if the assertion (verify phase) is too quick and value has not been updated. That's why we have these dynamic
> wait. The logic behind this method is that every `pollIntervalMs` for specific period of time `timeoutMs` the method checks

Reviewer: ```suggestion
can occur if the assertion (verify phase) is too quick and value has not been updated. That is the reason why we have this dynamic `waitFor()` function. The logic behind this method is that every `pollIntervalMs` for specific period of time `timeoutMs` the method checks
```

`BooleanSupplier` condition which user can provide and if this condition is true the method will successfully end. Otherwise, we will get the

> `WaitException` with some reasonable description. In case of interest method is located [here](https://github.com/strimzi/strimzi-kafka-operator/blob/master/test/src/main/java/io/strimzi/test/TestUtils.java#L115-L145).

Reviewer: ```suggestion
`WaitException` with some reasonable description. The method implementation can be found [here](https://github.com/strimzi/strimzi-kafka-operator/blob/master/test/src/main/java/io/strimzi/test/TestUtils.java#L115-L145).
```

#### Constants

> Purpose of this class is to keep global constants used across system tests. You can find the constants for timeout
> interval, poll internal used inside `waitFor` method. Moreover, you can find timeout for cluster operator and clients.
> Also, we have here specific ports such as KafkaBridge metric, TopicOperator metric, Keycloak and so on. Lastly, here we have all tags names,
> which we used in whole `system test` module for instance `ACCEPTANCE`, `REGRESSION` and more.

Reviewer: ```suggestion
Purpose of this class is to keep global constants used across system tests in one place. You can find there mentioned constants like timeout interval and poll interval used by `waitFor()` method. Also, there are various other timeouts for cluster operator and clients, definitions of specific ports such as KafkaBridge metric, TopicOperator metric, Keycloak and so on. Lastly, there are all tag names,
which are used in whole `system test` module, for instance `ACCEPTANCE`, `REGRESSION` and more.
```

 
#### Environment

System tests can be configured by several environment variables, which are loaded before test execution.

> Variables can be defined via environmental variables or a configuration file, this file can be located anywhere on the file system as long as a path is provided to this file.

Reviewer: ```suggestion
Variables can be defined via environmental variables or a configuration file. This file can be located anywhere on the file system as long as a path is provided to this file.
```

> The path is defined by environmental variable `ST_CONFIG_PATH`, if the `ST_CONFIG_PATH` environmental variable is not defined, the default config file location is used `systemtest/config.json`.

Reviewer: ```suggestion
The path is defined by environment variable `ST_CONFIG_PATH`. If it is not defined, the default config file location `systemtest/config.json` will be used instead.
```

Loading of system configuration has the following priority order:
1. Environment variable
2. Variable defined in configuration file
3. Default value

Here is the list of important environment variables:

> 1. `DOCKER_ORG` = Specify the organization/repo containing the image used in system tests (default value `strimzi`)
> 2. `DOCKER_TAG` = Specify the image tags used in system tests (default value `latest`)
> 3. `SKIP_TEARDOWN` = Variable for skip teardown phase for more debug if needed (default value `false`)
> 4. `ST_KAFKA_VERSION` =  Kafka version used in images during the system tests (default value `2.6.0`)

Reviewer: ```suggestion
1. `DOCKER_ORG` = Specifies the organization/repo containing the image used in system tests (default value `strimzi`)
2. `DOCKER_TAG` = Specifies the image tags used in system tests (default value `latest`)
3. `SKIP_TEARDOWN` = Skip teardown phase - primarily used for debug purposes (default value `false`)
4. `ST_KAFKA_VERSION` =  Specifies Kafka version used in images during the system tests (default value `2.6.0`)
```

#### Our Kafka clients

> Another helpful part of the system tests are Kafka clients. In this module we have three types:

Reviewer: ```suggestion
Another helpful module of the system tests are Kafka clients, which contains three usage types:
```

> 1. Internal using test-client image

Reviewer: Internal and external to what?

Something like `For internal communication within Kubernetes - using test-client image`?

2. Internal using client-examples image
3. External using Vert.x client

##### Internal using test-client image

The first client using the our image. Basically, it is defined as a deployment and inside the container, we start the 

> `kafka-verifiable-(consumer|producer).sh` shell script. In case of more interest you can [here](https://github.com/strimzi/strimzi-kafka-operator/tree/master/docker-images/test-client).

Reviewer 5: ```suggestion
`kafka-verifiable-(consumer|producer).sh` shell script. In case of more interest you can [here](https://github.com/strimzi/strimzi-kafka-operator/tree/release-0.20.x/docker-images/test-client).
```

For instance, if we want to create some producer or consumer we can easily do:

```java
// 1. Creating the client instance 

InternalKafkaClient internalKafkaClient = new InternalKafkaClient.Builder()
    .withUsingPodName("kafka-client-pod")
    .withTopicName("topic-name")
    .withNamespaceName("namespace-name")
    .withClusterName("cluster-name")
    .withMessageCount(100)
    .build();

// 2. Sending messages by interface method

internalKafkaClient.sendMessagesPlain();  <- for plain communication
```

Note, also that before sending messages to the Kafka cluster we need ensure that test-client pod is running and it can 
done via this code, which basically creates the pod,
where inside this one is running test-client and waiting for commands invoked by `kafka-verifiable-(consumer|producer).sh`.

```java
KafkaClientsResource.deployKafkaClients(false, CLUSTER_NAME + "-" + Constants.KAFKA_CLIENTS).done();
```

##### Internal using client-examples image

The second one is based on client-examples images, where everything about client implementation with also `Dockerfiles`
can be found [here](https://github.com/strimzi/client-examples).

> These type of client we are using as Job. For instance if we want to send some messages we write it like this:

Reviewer 5: URL with code where it is stored in Strimzi oeprator repo?

                                .withValue(String.valueOf(messageCount))
                            .endEnv()
                            .addNewEnv()
                                .withName("PRODUCER_ACKS")
                                .withValue("all")
                            .endEnv()
                            .addNewEnv()
                                .withName("ADDITIONAL_CONFIG")
                                .withValue(additionalConfig)
                            .endEnv()
                    .endContainer()
                .endSpec()
            .endTemplate()
        .endSpec()
        .build());
```

##### External using Vert.x client

Lastly, we have our external clients implemented by support of Vertx.io Verticles. This type of client has the most limitation such as
dependent on external listeners, which caused `platform dependency`. 
Implemented by Abstract recursive builder, which makes these clients very handful. 
Moreover, we have `Oauth`, `Tracing` and also `Basic` client. Each of them are for specific use case. 
However, we prefer to always use the internal clients because of that platform dependency. 

> Client can be used for instance:

Reviewer 5: Maybe add note that external clients are running where tests are running, and internal inside the kube cluster

```java
// 1. Creating the client instance 

BasicExternalKafkaClient basicExternalKafkaClient = new BasicExternalKafkaClient.Builder()
    .withTopicName("topic-name")
    .withNamespaceName("namespace-name")
    .withClusterName("cluster-name")
    .withKafkaUsername("kafka-username")
    .withMessageCount(100)
    .withSecurityProtocol(SecurityProtocol.SSL)
    .build();

// 2. Sending messages

basicExternalKafkaClient.sendMesssageTls();
```

## Dependencies

> Before we get to writing simple test case in Strimzi system tests, we need to know main dependencies that `system test` module have. 
> I gonna describe the most important one like:
> 1. test (kubernetes client)
> 2. crd-generator + api (chaining / builder pattern)

Reviewer: ```suggestion
Before writing simple test case in Strimzi system tests, we need to know main dependencies that `system test` module has. 
The most important ones are:
1. test (kubernetes client)
2. crd-generator + api (chaining / builder pattern)
```

### test module

In the test module, we have the Kubernetes client, which we use in the system tests to create, list, delete, update
Kubernetes resources such as Service, Deployment, Pod, ConfigMap and so on. Basically, we encapsulate the [Kubernetes client](https://github.com/fabric8io/kubernetes-client)
from fabric8 to use client more handy with not additional and typically redundant method chaining. 
Moreover, we have our abstraction of Kubernetes cluster and mechanism to detect on which type is currently client. 

> Clusters could be Kubernetes, Minikube and so on.

Reviewer 5: Missing info about KubeCluster representation and cmdClient

> In the test module, we have the Kubernetes client, which we use in the system tests to create, list, delete, update
> Kubernetes resources such as Service, Deployment, Pod, ConfigMap and so on. Basically, we encapsulate the [Kubernetes client](https://github.com/fabric8io/kubernetes-client)
> from fabric8 to use client more handy with not additional and typically redundant method chaining. 
> Moreover, we have our abstraction of Kubernetes cluster and mechanism to detect on which type is currently client. 
> Clusters could be Kubernetes, Minikube and so on.

Reviewer: ```suggestion
In the test module, we have the Kubernetes client, which we use in the system tests to create, list, delete and update Kubernetes resources such as Service, Deployment, Pod, ConfigMap and so on. Basically, we encapsulate the [Kubernetes client](https://github.com/fabric8io/kubernetes-client)
from fabric8 to use client more handy with not additional and typically redundant method chaining. 
Moreover, we have our abstraction of Kubernetes cluster and mechanism to detect on which type is currently client. 
Clusters could be Kubernetes, Minikube and so on.
```

### crd-generator + api module (not sure about this one...)

> In the crd-generator and api modules, we basically use all resources, which we previously described in Resource section. 
> Every time, if the code of Model changes you have to build the whole modules to generate new schema for resources. 
> For instance imagine that you are using `Kafka` resource and you need to create Kafka with the `Cruise Control` 
> support but you have got old schemas, which does not support it.
> You will need to build these modules and generate schema, which will support it and be able to create Kafka with Cruise Control. 
> This modules provides us the fluent way of creating Strimzi custom resources instances such as `Kafka`, `KafkaConnect`, 
> `KafkaConnectS2I`, `KafkaMirrorMaker` and much more.

Reviewer: ```suggestion
Resources previously described in Resource section are used in crd-generator and api modules.
Every time, when the code of Model changes you have to build the whole modules to generate new schema for resources. 
For instance, imagine that you are using `Kafka` resource and you need to create Kafka with the `Cruise Control` support but you have old schemas, which does not support it.
You will need to build these modules and generate schema, which will support it and be able to create Kafka with Cruise Control. 
These modules provide the fluent way of creating Strimzi custom resource instances such as `Kafka`, `KafkaConnect`, `KafkaConnectS2I`, `KafkaMirrorMaker` and many other.
```

## How to create a system test

> Now, we have all knowledge to create a simple test. There is also some precursors, which i will write it down.

Reviewer: ```suggestion
At this point you have the necessary knowledge to create a simple test. There are also some additional steps, which need to be performed to set up your testing environment.
```

### Precursors

1. Fork the `strimzi-kafka-operator` repository - https://github.com/strimzi/strimzi-kafka-operator

> 2. Build the whole project in root directory - `mvn clean install -DskipTests`

Reviewer 5: Or you can use `mvn -am clean install -DskipTests -f systemtest/pom.xml`

3. Setup your Kubernetes cluster
4. Login to your Kubernetes cluster

> 5. Try to start some of our tests inside `system test` module for instance - `io.strimzi.systemtest.bridge.HttpBridgeST.testSendSimpleMessage()` 

Reviewer 5: Shouldn't be pick differet scenario? Like `testSendMessagesTlsAuthenticated`

Author: Ok :)

> (take takes around 5 minutes)

Reviewer: ```suggestion
(test takes around 5 minutes)
```

### Writing new test case

> We will try to create a very simple test using our `Resources`, `Utils`, `Kubernetes client` and so on.
> 
> Description of the test case: We need to verify that Kafka can be deployed without any problem and then that
> clients can send messages to Kafka cluster and receive messages from them. 
> So as you can see, we have two main things to take care. 
> First is `verify that Kafka can be deployed` and the second one is `clients can send messages to Kafka cluster and receive messages from them`.

Reviewer: ```suggestion
We are going to create a very simple test using aforementioned classes like `Resources`, `Utils`, `Kubernetes client` and so on.

Description of the test case: 

1. Verify that Kafka was successfully deployed
2. Clients are able to send and receive messages to Kafka cluster
 
As you can see, we have two main things to take care of. First is `verify that Kafka was correctly deployed` and the second one is `clients can send messages to and receive messages from Kafka cluster`.
```

Test scenario could look like this:

1. Deploy Strimzi Operator <- setup + exercise
2. Deploy `Kafka` cluster with 3 kafka nodes and 3 zookeeper nodes <- setup + exercise

> 3. Verify that `Kafka` is ready <- verify

Reviewer 5: This is basically done automatically, maybe you should mention it

4. Create instance of `Kafka clients` <- setup + exercise
5. Send messages to `Kafka` cluster <- exercise
6. Receive messages from `Kafka` cluster <- exercise

> 7. Verify that we have sent and received excepted count of messages <-verify

Reviewer: ```suggestion
7. Verify that we have sent and received expected count of messages <-verify
```

8. Clean up all resources (this phase is implicitly done in AbstractST) <- teardown

> Note, that teardown phase is down automatically done by our mechanism so there is no need to take care of it.

Reviewer: ```suggestion
Note, that teardown phase is performed automatically by system tests mechanism so there is no need to take care of it by user in code.
```

        String defaultKafkaClientsPodName =
            ResourceManager.kubeClient().listPodsByPrefixInName(CLUSTER_NAME + "-" + Constants.KAFKA_CLIENTS).get(0).getMetadata().getName();

        InternalKafkaClient internalKafkaClient = new InternalKafkaClient.Builder()
            .withUsingPodName(defaultKafkaClientsPodName)
            .withTopicName(TOPIC_NAME)
            .withNamespaceName(NAMESPACE)
            .withClusterName(CLUSTER_NAME)
            .withMessageCount(MESSAGE_COUNT)
            .build();

        // 4.point ends
        // ============

        LOGGER.info("Checking produced and consumed messages to pod:{}", defaultKafkaClientsPodName);

        internalKafkaClient.checkProducedAndConsumedMessages(  // <-- 7.point
            internalKafkaClient.sendMessagesPlain(),   // <-- 5.point
            internalKafkaClient.receiveMessagesPlain() // <-- 6.point
        );
    }

    @BeforeAll
    void setup() throws Exception {
        ResourceManager.setClassResources();
        installClusterOperator(NAMESPACE); //  <-- 1.point
    }
}

```

Congratulations you have written your new test case in Strimzi `system test` module!

## Conclusion

> The whole ecosystem of system tests is at first glance hard to understand but if the person who is learning this domain
> got a drive, It won't be a challenge for him. 
> Whole module passed through several changes, which improve many areas.
> We are open to  any discussion, which can create or change the behavior to robust, better way of tests.
> Maybe you will be the one who will change the structure of system tests and make it even better!

Reviewer: ```suggestion
The whole ecosystem of system tests is at first glance hard to understand but if the person who is learning this domain has a drive, it won't be a challenge for him. 
Whole module went through several iterations of evolution, which improved many areas.
We are open to any discussion and suggestions, which can enhance the overall behavior to robust, better way of testing.
Maybe You will be the one who will change the structure of system tests and make it even better!
```

---
layout: post
title:  "How system tests works"
date: 2020-09-06

> author: maroš_orsák

Reviewer 2: I think you should fix the name to the ones in the authors file.

---

Sometimes, we underestimate strength of testing. Many companies start verification of the software product belated.
The process of testing is complex as many can not imagine. 
It can be highlighted by quote: "If we fail, we fall. If we succeed - then we will face the next task." 
Fortunately, Strimzi is an exception but you have to know many technologies such as [Kubernetes](https://kubernetes.io/), 
[Apache Kafka](https://kafka.apache.org/),[Keycloak](https://www.keycloak.org/), [Open Policy Agent](https://www.openpolicyagent.org/),

> [Cruise control](https://github.com/linkedin/cruise-control) and so on.

Reviewer 2: I absolutely don't understan how this fits in.

> In this blog post we are gonna take a closer look on our system tests. The content is as follows:

Reviewer 2: I think in here you need to link to the first blog post and somehow follow up on it to make it more coherent.

# Content
1. [Introduction](#Introduction)
2. [Resources](#Resources)
3. [Lifecycle of tests](#Lifecycle of tests)
4. [Auxiliary classes](#Auxiliary classes)
5. [Dependencies](#Dependencies)
6. [How to create a system test](#How to create a system test)
8. [Conclusion](#Conclusion)

## Introduction

The whole ecosystem behind systems test is encapsulated in two mandatory things. First of all it is

> the Resources, which create the whole testing environment. The second of all we have the auxiliary classes that are divided into

Reviewer 4: maybe like this?:
```suggestion
the Resources, which create the whole testing environment. The second of all we have the auxiliary classes that are divided into
```
```suggestion
the `Resources`, which create the whole testing environment. The second of all we have the auxiliary classes that are divided into
```

the classic static methods `Utils`, internal clients and also `Apache Kafka Clients` for the external communication, 
Kubernetes client, `Constants.java` and `Environment.java`.

Pure idea to make the system tests easily modifiable is to able to write resources in the fluent way. 
Fluent notation are nested builders by which you can build customized objects by your imagination.

> For instance before the Fluent notation we have the [1.Telescoping pattern](https://www.vojtechruzicka.com/avoid-telescoping-constructor-pattern/#telescoping-constructor) 

Reviewer 4: Maybe it's just my bad english but:
```suggestion
For instance before the Fluent notation we had the [1.Telescoping pattern](https://www.vojtechruzicka.com/avoid-telescoping-constructor-pattern/#telescoping-constructor) 
```

but this does not scale well for instance in situation with large numbers of optional parameters. 
Second alternative would be the [2.JavaBeans convention](https://www.vojtechruzicka.com/avoid-telescoping-constructor-pattern/#alternative-1---javabeans), 
which first of all resolve our issue from the telescoping pattern but on the other hand the objects which we create will be mutable. 
Luckily, there is a third alternative that combines the safety of the telescoping constructor pattern with the readability 
of the `JavaBeans pattern`. 
It is a form of the [3.Builder pattern](https://refactoring.guru/design-patterns/builder).
Moreover, if we take this idea to the next level to the `4.Abstract Recursive builder` we will be able to create multiple 

> dependent objects, which have nested hierarchy. In the following code snippets 1.1, 1.2, 1.3 we can see described idea and you 

Reviewer 4: I think that you forgot to change the snippets "names" down there. Also is it possible to add some reference to the snippet? Just an idea..

can see how easily we can create an object using `Abstract Recursive builder`.

####  1. Kafka with new persistent claim storage and following settings

##### Code snippet 1.2 Using JavaBeans convention
```java
PersistentClaimStorage psc = new PersistentClaimStorage();

// JavaBeanss convention having setters 
psc.setId(0);
psc.setSize("100Gi");
psc.setDeleteClaim(true);

KafkaResource.kafkaEphemeral(CLUSTER_NAME, 3, 1)
    .editSpec()
        .editKafka()
            .withPersistentClaimStorage(psc)
        .endKafka()
    .endSpec()
    .done();
```

##### Code snippet 1.3 Using Builder pattern
```java
// Builder pattern by which we can create one type of object
PersistentClaimStorage pcs = new PersistentClaimStorageBuilder()
    .withDeleteClaim(true)
    .withId(0)
    .withSize("100Gi")
    .build();

KafkaResource.kafkaEphemeral(CLUSTER_NAME, 3, 1)
    .editSpec()
        .editKafka()
            .withPersistentClaimStorage(psc)
        .endKafka()
    .endSpec()
    .done();
```

> ##### Code snippet 1.4 Using Abstract recursive builder

Reviewer 4: Names/ordering how I mentioned above

        .withName(name)
        .withNamespace(ResourceManager.kubeClient().getNamespace())
    .endMetadata()
    .editSpec()
        .editKafka()
            .withVersion(Environment.ST_KAFKA_VERSION)
            .withReplicas(kafkaReplicas)
            .addToConfig("log.message.format.version", TestKafkaVersion.getKafkaVersionsInMap().get(Environment.ST_KAFKA_VERSION).protocolVersion())
            .addToConfig("offsets.topic.replication.factor", Math.min(kafkaReplicas, 3))
            .addToConfig("transaction.state.log.min.isr", Math.min(kafkaReplicas, 2))
            .addToConfig("transaction.state.log.replication.factor", Math.min(kafkaReplicas, 3))
            .withNewListeners()
                .withNewPlain().endPlain()
                .withNewTls().endTls()
            .endListeners()
            .withNewInlineLogging()
                .addToLoggers("log4j.rootLogger", "DEBUG")
            .endInlineLogging()
        .endKafka()
        ...
        ... (more configs...)
        ... 
    .endSpec()
    .done()
```

In this case you can imagine that it is really easy to create multiple different instances of the `Kafka` instead of having separate objects.

> ## Resources

Reviewer 4: Can you add the `2.` so the figures will be correct?

Author: Hmm, then It would be not so great and I will need to update all content titles. Resources is another chapter and `1. Kafka with new persistent claim storage and following settings` was just additional information in the Introduction chapter.

Resources in our way are abstraction to the `.yaml` definitions of custom resources which we provide such as 
`Kafka`, `KafkaTopic`, `KafkaUser`, `KafkaConnect`, `KafkaMirrorMaker` and so on.
These custom resources are encapsulated into java objects and some of you can remind the [ORM(Object Relation Mapping)](https://www.tutorialspoint.com/hibernate/orm_overview.htm) 
from databases. 
In the following Figure 2.1 we can see YAML representation of `KafkaTopic` and on the Code snippet 2.2 related java object.

> Figure 2.1 YAML representation of `KafkaTopic`

Reviewer 4: Maybe some header? Like it is in code snippets above?

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: KafkaTopic
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec:
  partitions: 1
  replicas: 1
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824
```

> Code snippet 2.2 Loading the YAML file to JAVA representation object.

Reviewer 4: Same as above

#### Constants

Purpose of this class is to keep global constants used across system tests. You can find the constants for timeout
interval, poll internal used inside `waitFor` method. Moreover, you can find timeout for cluster operator and clients.
Also, we have here specific ports such as KafkaBridge metric, TopicOperator metric, Keycloak and so on. Lastly, here we have all tags names,
which we used in whole `system test` module for instance `ACCEPTANCE`, `REGRESSION` and more.
 
#### Environment

System tests can be configured by several environment variables, which are loaded before test execution.

Variables can be defined via environmental variables or a configuration file, this file can be located anywhere on the file system as long as a path is provided to this file.
The path is defined by environmental variable `ST_CONFIG_PATH`, if the `ST_CONFIG_PATH` environmental variable is not defined, the default config file location is used `systemtest/config.json`.
Loading of system configuration has the following priority order:
1. Environment variable
2. Variable defined in configuration file
3. Default value

Here is the list of important environment variables:

> 1. DOCKER_ORG = Specify the organization/repo containing the image used in system tests (default value `strimzi`)
> 2. DOCKER_TAG = Specify the image tags used in system tests (default value `latest`)
> 3. SKIP_TEARDOWN = Variable for skip teardown phase for more debug if needed (default value `false`)
> 4. ST_KAFKA_VERSION =  Kafka version used in images during the system tests (default value `2.6.0`)

Reviewer 4: ```suggestion
1. `DOCKER_ORG` = Specify the organization/repo containing the image used in system tests (default value `strimzi`)
2. `DOCKER_TAG` = Specify the image tags used in system tests (default value `latest`)
3. `SKIP_TEARDOWN` = Variable for skip teardown phase for more debug if needed (default value `false`)
4. `ST_KAFKA_VERSION` =  Kafka version used in images during the system tests (default value `2.6.0`)
```

about our unit tests, integration tests and how they differ from system tests. 
Furthermore, we showed up how to run them on Kubernetes cluster. 
But, we did not cover whole understanding of the system tests. 
In this blog post we will take a closer look at our system tests. The content is as follows:

# Content
1. [Introduction](#Introduction)
2. [Resources](#Resources)
3. [Lifecycle of tests](#Lifecycle of tests)
4. [Auxiliary classes](#Auxiliary classes)
5. [Dependencies](#Dependencies)
6. [How to create a system test](#How to create a system test)
8. [Conclusion](#Conclusion)

## Introduction

The whole ecosystem behind systems test is encapsulated in two mandatory things. First the `Resources`, creates the whole 
testing environment. Second - the auxiliary classes, which are divided into the `Resources`, which create the whole 
testing environment. 
The second of all we have the auxiliary classes that are divided into
the classic static methods `Utils`, internal clients and also `Apache Kafka Clients` for the external communication, 
Kubernetes client, `Constants.java` and `Environment.java`.

> Main idea was to make the system tests and its resources easily modifiable and writable in the fluent way. 

Author: is :)

Fluent notation is a set of nested builders, which can build customized objects by your needs.
For instance before the Fluent notation we had the [1.Telescoping pattern](https://www.vojtechruzicka.com/avoid-telescoping-constructor-pattern/#telescoping-constructor) 
but that did not scale well (for instance in situations with large numbers of optional parameters). 
Second alternative would be the [2.JavaBeans convention](https://www.vojtechruzicka.com/avoid-telescoping-constructor-pattern/#alternative-1---javabeans), 
which could resolve our issue from the telescoping pattern, but on the other hand objects which created objects will be mutable. 
Luckily, there is a third alternative that combines the safety of the telescoping constructor pattern with the readability 
of the `JavaBeans pattern`. 
It is a form of the [3.Builder pattern](https://refactoring.guru/design-patterns/builder).
Moreover, if we take this idea to the next level to the `4.Abstract Recursive builder` we will be able to create multiple 
dependent objects, which have nested hierarchy. In the following code snippets [1.1](#javabeans), [1.2](#builder), [1.3](#abstractbuilder) 

> we can see described idea and you can see how easily we can create an object using `Abstract Recursive builder`. 

Author: I have got your points and I removed all related things about patterns. At the start, I was thinking that this may fit here but when I read it again it was not so great to me like it was for the first time.

from databases. 
In the following yaml code[2.1](#kafkatopic) we can see YAML representation of `KafkaTopic` and on the Code snippet 
[2.2](kafkaimplementation) related java object.

##### <a id="kafkatopic">Code snippet 2.1 YAML representation of `KafkaTopic`</a>
```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: KafkaTopic
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec:
  partitions: 1
  replicas: 1
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824
```

##### <a id="kafkaimplementation">Code snippet 2.2 Loading of the YAML file to JAVA representation object.</a>
```java
KafkaTopic kafkaTopic = getKafkaTopicFromYaml(PATH_TO_KAFKA_TOPIC_CONFIG); 
// additional changes...
kafkaTopic = new KafkaTopicBuilder(kafkaTopic)
    .withNewMetadata()
        .withName(topicName)
        .withNamespace(ResourceManager.kubeClient().getNamespace())
        .addToLabels(Labels.STRIMZI_CLUSTER_LABEL, clusterName)
    .endMetadata()
    .editSpec()
        .withPartitions(partitions)
        .withReplicas(replicas)
        .addToConfig("min.insync.replicas", minIsr)
    .endSpec();
```

> Everything is built within three stacks. These three stacks are managed by `ResourceManager`, which is taking care of switching. 

Author: I have added link to these stacks our code.

First one's, `method stack`, responsibility is to hold all resources, which are invoked in the test cases. 
Second one's, the `class stack`, responsibility is to hold all resources, which are invoked in the `@BeforeAll` notation. 
The third stack is used as a pointer to either class or method stack to delete proper resources in specific test phases.
The `pointer stack` points to the `class stack` in we are in the `@BeforeAll` scope, otherwise we are working with the `method stack`. 
The logic inside these stacks are that once you create some resource (for instance KafkaTopic),
it will be pushed inside the stack and the deletion of topic will be performed in the teardown phase. 
Once the teardown phase is initiated, the `method stack` will pop all resources after test case is over. The same logic applies with `class stack,
which will pop all resources when test suite is over. 

> What is worth to mention, is that if you specify `SKIP_TEARDOWN` environment variable to `TRUE` it will skip the teardown 

Author: Thanks for the commend. Are you suggesting to put `## Lifecycle of tests` as a second chapter?

> What is worth to mention, is that if you specify `SKIP_TEARDOWN` environment variable to `TRUE` it will skip the teardown 

Reviewer 2: Yeah, I guess that was the thought I had when writing the comment.

> What is worth to mention, is that if you specify `SKIP_TEARDOWN` environment variable to `TRUE` it will skip the teardown 

Author: Done :)

phase and all created resources will remain in the stack.

## Lifecycle of tests

Like any other test life cycle, we follow the [SEVT](http://xunitpatterns.com/Four%20Phase%20Test.html) convention. 
[SEVT](http://xunitpatterns.com/Four%20Phase%20Test.html) stands for Setup, Exercise, Verify and Teardown.
The fundamental idea can be illustrated as follows:

```
1.SETUP        | instance = new Instance(); 
2.EXERCISE     | instance.setSomeValue(someValue);
3.VERIFY       | asserThat(instance.getSomeValue(), is(someValue));
4.TEARDOWN    | instance = null; (this steps is prevailingly done by testing Frameworks)

> ```

Author: removed

#### Setup

In `Setup phase` we perform following actions:

* Create namespace(s)
* Deploy the Strimzi Cluster operator
* Deploy the Kafka cluster and/or other components (optional)

> The reason why the last point is optional, is because we have some test cases where we might want to have a different kafka 

Reviewer 2: Following the table above, I'm not sure this is true because from there it does not sound like setup is only the `@BeforeClass`. It could be also the first part of the test method.

```java
    @Override
    protected void tearDownEnvironmentAfterAll() {
        doSomethingYouNeed();
        super.tearDownEnvironmentAfterAll();
    }
```

In order to delete all resources from specific `Resources` instance, execute:
```java
    ResourceManager.deleteMethodResources();
    ResourceManager.deleteClassResources();
```

## Auxiliary classes

In `system test` module there are many helper classes,  each having a different purpose:

1. Utils
2. Constants
3. Environment
4. Kubernetes client
5. Kafka clients

#### Utils

These classes are applicable on various problems. We have separate utils class for each resource. 
For instance we have KafkaUtils for KafkaResource, KafkaTopic for KafkaResource and so on. Furthermore, we have got
Kubernetes related utils such as ServiceUtils, PodUtils, DeploymentsUtils and many more. The hierarchy is as follows:

* utils
  * kafkaUtils
    * KafkaBridgeUtils
    * KafkaConnectS2IUtils
    * ...
  * kubeUtils
    * controllers
      * DeploymentConfigUtils
      * DeploymentUtils
      * ...
    * objects
      * NamespaceUtils
      * PersistentVolumeClaimUtils
      * ...
  * specific
    * BridgeUtils

>     * CruiseControlUtils

Reviewer 2: What is this structure? packages and classes? Is it really needed when you anyway, always end up with `...`?

Author: removed

---
layout: post
title:  "How the SYSTEM TESTS works"
date: 2020-09-06
author: maroš_orsák
---

> Sometimes, we underestimate strength of testing. Many companies start verification of the software product belated. The process of testing is complex as many can not imagine. It can be highlighted by quote: "If we fail, we fall. If we succeed - then we will face the next task."

Reviewer 2: You should start with something more optimistic and not how bad we are at system tests :-o

the classic static methods `Utils`, Clients for the external and also internal communication, Kubernetes client, `Constants.java` and `Envinroment.java`.

Pure idea to make the system tests easily modifiable is to able to write resources in the fluent way. Fluent notation is nested builders by which you can build customized objects by your imagination.
For instance before the Fluent notation we have the [1.Telescoping pattern](https://www.vojtechruzicka.com/avoid-telescoping-constructor-pattern/#telescoping-constructor) but this does not scale well to large numbers of optional parameters.
Second alternative would be the [2.JavaBeans convention](https://www.vojtechruzicka.com/avoid-telescoping-constructor-pattern/#alternative-1---javabeans), which first of all resolve our issue from the telescoping pattern but on the
other hand the objects which we create will be mutable. Luckily, there is a third alternative that combines the safety of the telescoping constructor pattern with the readability of the JavaBeans pattern. It is a form of the [3.Builder pattern](https://refactoring.guru/design-patterns/builder).

> Moreover, if we take this idea to the next level to the `4.Abstract Recursive builder` we will be able to create multiple dependent objects, which has some hierarchy. In the following figures 1.1, 1.2, 1.3 we can see described idea and you can see how easily we can create an object using Abstract Recursive builder.

Reviewer 6: they are code snippets, not figures

            .withReplicas(kafkaReplicas)
            .addToConfig("log.message.format.version", TestKafkaVersion.getKafkaVersionsInMap().get(Environment.ST_KAFKA_VERSION).protocolVersion())
            .addToConfig("offsets.topic.replication.factor", Math.min(kafkaReplicas, 3))
            .addToConfig("transaction.state.log.min.isr", Math.min(kafkaReplicas, 2))
            .addToConfig("transaction.state.log.replication.factor", Math.min(kafkaReplicas, 3))
            .withNewListeners()
                .withNewPlain().endPlain()
                .withNewTls().endTls()
            .endListeners()
            .withNewInlineLogging()
                .addToLoggers("log4j.rootLogger", "DEBUG")
            .endInlineLogging()
        .endKafka()
        ...
        ... (more configs...)
        ... 
    .endSpec()
    .done()
```

In this case you can imagine that it is really easy to create multiple different instances of the Kafka instead of having separate objects.

## Resources

Resources in our way are abstraction to the `.yaml` definitions of custom resources which we provide such as Kafka, KafkaTopic, KafkaUser, KafkaConnect, KafkaMirrorMaker and so on.

> These custom resources are encapsulated into java objects and some of you can remind the [ORM(Object Relation Mapping)](https://www.tutorialspoint.com/hibernate/orm_overview.htm) from databases. In the following Figure 1.1 we can see YAML representation of KafkaTopic and on the Figure 1.2 related JAVA object.

Reviewer 6: 2.1 not 1.1, 2.2 not 1.2

        .addToConfig("min.insync.replicas", minIsr)
    .endSpec();
```

Everything is build with three stacks. These three stacks are managed by `ResourceManager`, which is taking care of switching. First one is `method stack` where his responsibility is to hold all
resources, which are invoked in the test cases. Second one is the `class stack` where his responsibility is to hold all resources,
which are invoked in the @BeforeAll notation. Switching between these two stacks are doing the third stack called pointer stack. The pointer stack points to class stack where it is in the @BeforeAll notation otherwise we are working with the method stack. The logic inside these stacks are that if you create some resource
for instance KafkaTopic. It will be pushed inside the stack and the deletion of topic will be planned. In the teardown phase method stack will pop all resources after test case is over and same logic applied with class stack, which will pop all resources when test suite is over. What is worth mentioning is that if you specify `SKIP_TEARDOWN` environment variable to `TRUE` it will skip the teardown phase and all created resources will be there.

## Lifecycle of tests

> Like any other test life cycle, ours are no different and follow the SEVT convention. SEVT stands for Setup, Exercise, Verify and Teardown.

Reviewer 6: is there any link to this SEVT convention?

Author: http://xunitpatterns.com/Four%20Phase%20Test.html Yes I will also put it there.

      * ReplicaSetUtils
      * StatefulSetUtils
    * objects
      * NamespaceUtils
      * PersistentVolumeClaimUtils
      * PodUtils
      * SecretUtils
      * ServiceUtils
  * specific
    * BridgeUtils
    * CruiseControlUtils
    * KeycloakUtils
    * MetricUtils
    * TracingUtils
    
The usability of these classes is for instance when you have to to wait for some change, calling API, creating, changing, deleting Kubernetes resource and much more.
The one method which deserves to be described is `waitFor`, where you will find it in many utils methods.
You can imagine the scenario, where you will change some custom resource value of Kafka and then you will want to assert
that this value was updated but here is very good example where comes `race condition` into play. This race condition
can occur if the assertion (verify phase) is too quick and value has not been updated. That's why we have these dynamic
wait. The logic behind this method is that every `pollIntervalMs` for specific period of time `timeoutMs` the method checks
`BooleanSupplier` condition which user can provide and if this condition is true the method will successfully end. Otherwise, we will get the

> `WaitException` with some reasonable description. In the Figure 4.1 we can see the whole `waitFor` method.

Reviewer 6: I don't know how much is really useful posting the whole code for the method. Just describing its purpose, the parameters meaning should be ok. If the reader wants, he can just go to the source code in the repo (maybe you can add the link here).

Author: Make sense :)

                                        .withValue("1000")
                                    .endEnv()
                                    .addNewEnv()
                                        .withName("LOG_LEVEL")
                                        .withValue("DEBUG")
                                    .endEnv()
                                    .addNewEnv()
                                        .withName("MESSAGE_COUNT")
                                        .withValue(String.valueOf(messageCount))
                                    .endEnv()
                                    .addNewEnv()
                                        .withName("PRODUCER_ACKS")
                                        .withValue("all")
                                    .endEnv()
                                    .addNewEnv()
                                        .withName("ADDITIONAL_CONFIG")
                                        .withValue(additionalConfig)
                                    .endEnv()
                            .endContainer()
                        .endSpec()
                    .endTemplate()
                .endSpec()
                .build());
```

> ##### External using Vertx.io client

Reviewer 6: It's Vert.x not Vertx.io :-)
